### ISSUE TITLE
Cross-thread Queue

### ISSUE THREAD

**Issue by andrew-d** [saghul/evergreen#8]
Hi there,

I'm using Evergreen in [a project of mine](https://github.com/andrew-d/holepunch/), and I'm running into a strange edge case.  Long story short, I need to read from a TUN/TAP device on Mac OS X.  If you do pretty much anything except calling select() on the device, you get errno = 35.  So, I tried calling select() from a regular Thread, reading the data that shows up, and then putting it on a queue (see [here](https://github.com/andrew-d/holepunch/blob/66093be7598d1a79a90340b1cb14ab23c26ced64/holepunch/tuntap/darwin.py#L18)).  However, I get an `RuntimeError('there is no event loop created in the current thread')`, which makes sense.  However, I don't know of any other way to do this - sending data to a task from another thread without blocking the entire event loop.  Any suggestions?

Thanks!

**Comment by saghul** [comment]
Hi!

Evergreen uses pyuv (Python bindings to libuv, nodejs core) internally and I know this case is internally handled, so you should be able to use a Pipe handle and read from there. I don't have an API for making this easy just yet, but you can try to use the pyuv API for the time being.

Get the fd, and open it with a Pipe handle (https://pyuv.readthedocs.org/en/release-0.10.0/pipe.html). Then you can call start_read and you'll get a callback (called on the same thread) with the data. Rough pseudocode:

```
def read_cb(handle, data, errno):
    print(args)

loop = evergreen.EventLoop()

name = 'tun0'
fd = os.open('/dev/' + name, os.O_RDWR)
pipe = pyuv.Pipe(loop._loop)
pipe.open(fd)
pipe.start_read(read_cb)

loop.run()
```

I plan to add the open API to do this easier using the io module, so it will be easier in the future, but you can use pyuv level APIs for now.

**Comment by saghul** [comment]
Alternatively, you can use the method you are using now, but instead of using a Queue use loop.call_from_thread to schedule a callback from the thread reading from the tun device to the thread where the loop is running. You can't use the evergreen queue across threads because it's cooperative, it works with tasks only.

**Comment by andrew-d** [comment]
Interestingly enough, the first method you posted (the Pipe one) gives me this error, followed by the process dying:

`Assertion failed: (events == (events & (UV__POLLIN | UV__POLLOUT))), function uv__stream_osx_select_cb, file src/unix/stream.c, line 248.`

Here's the exact code I'm using, on Mac OS X 10.7.5:

```
import evergreen, pyuv, os

def read_cb(*args):
    print(args)

loop = evergreen.EventLoop()

name = 'tun0'
fd = os.open('/dev/' + name, os.O_RDWR)

# E.g.
#   sudo ifconfig tun0 10.93.0.1 10.93.0.2 netmask 255.255.0.0
#   sudo route add -net 1.2.3.0 10.93.0.1 255.255.255.0
raw_input("configure the tun device, then press enter")

pipe = pyuv.Pipe(loop._loop)
pipe.open(fd)
pipe.start_read(read_cb)

loop.run()
```

**Comment by andrew-d** [comment]
After more investigation, looks like this is relevant:

joyent/node#5353

And looks like the fix is in joyent/libuv#791 and joyent/libuv#794.  I'm going to see if I can get pyuv building with that revision and see if it fixes the problem.

**Comment by andrew-d** [comment]
Ok, last comment for a while ;-) 

I've asked over at joyent/libuv#791 if they're planning on landing the fix(es) on the main branch of libuv, since I can't get pyuv compiling with the v0.10 branch (which is not entirely unexpected).  Until then, I'll go back to using threads, since any reasonable solution doesn't really work and just crashes on me.

**Comment by saghul** [comment]
It is unexpected. pyuv master targets libuv master, but pyuv branch v0.10 (current stable) targets libuv 0.10.x, so just modifying the commit hash here: https://github.com/saghul/pyuv/blob/v0.10/setup_libuv.py#L98 will do.

**Comment by schlamar** [comment]
And pass `--libuv-force-fetch` to setup.py so it actually downloads the new version and doesn't use the existing one :)

**Comment by saghul** [comment]
Indeed, thanks for pointing it out @schlamar :-)

**Comment by andrew-d** [comment]
Hokay, all good now :-)  I tried modifying the commit hash, but I was doing it on master, instead of the v0.10 branch (doh!), so it obviously didn't compile properly.  Just to follow up, the Pipe-based interface you posted above works fine with a TUN device on the commit hash `67f9b91`.  Thanks for putting up with me :-)

**Comment by saghul** [comment]
Great! I plan to release Evergreen 0.0.2 shortly, I'll see if I can squeeze in the API for this in the io module, here is an example of how it would look like:

```
from evergreen.io import PipeStream

pipe = PipeStream()
pipe.open(fd)

while True:
      data = pipe.read_bytes(4096)
      if not data:
          break
      # do smething
```

**Comment by saghul** [comment]
@andrew-d I just released pyuv 0.10.3 which includes those fixes from libuv.
